docs/docs/render-props.md

```diff
@@ -0,0 +1,224 @@
+  }
+
+  handleMouseMove(event) {
+    this.setState({
+      x: event.clientX,
+      y: event.clientY
+    });
+  }
+
+  render() {
+    return (
+      <div style={{ height: '100%' }} onMouseMove={this.handleMouseMove}>
+
+        {/* ...but how do we render something other than a <p>? */}
+        <p>The current mouse position is ({this.state.x}, {this.state.y})</p>
+      </div>
+    );
+  }
+}
+
+class MouseTracker extends React.Component {
+  render() {
+    return (
+      <div>
+        <h1>Move the mouse around!</h1>
+        <Mouse/>
+      </div>
+    );
+  }
+}
+```
+
+Now the `<Mouse>` component encapsulates all behavior associated with listening for `mousemove` events and storing the (x, y) position of the cursor, but it's not yet truly reusable.
+
+For example, let's say we have a `<Cat>` component that renders the image of a cat chasing the mouse around the screen. We might use a `<Cat mouse={{ x, y }}>` prop to tell the component the coordinates of the mouse so it knows where to position the image on the screen.
+
+As a first pass, you might try rendering the `<Cat>` *inside `<Mouse>`'s `render` method*, like this:
+
+```js
+class Cat extends React.Component {
+  render() {
+    const mouse = this.props.mouse
+    return (
+      <img src="/cat.jpg" style={{ position: 'absolute', left: mouse.x, top: mouse.y }}/>
+    );
+  }
+}
+
+class MouseWithCat extends React.Component {
+  constructor(props) {
+    super(props);
+    this.handleMouseMove = this.handleMouseMove.bind(this);
+    this.state = { x: 0, y: 0 };
+  }
+
+  handleMouseMove(event) {
+    this.setState({
+      x: event.clientX,
+      y: event.clientY
+    });
+  }
+
+  render() {
+    return (
+      <div style={{ height: '100%' }} onMouseMove={this.handleMouseMove}>
+
+        {/*
+          We could just swap out the <p> for a <Cat> here ... but then
+          we would need to create a separate <MouseWithSomethingElse>
+          component every time we need to use it, so <MouseWithCat>
+          isn't really reusable yet.
+        */}
+        <Cat mouse={this.state}/>
+      </div>
+    );
+  }
+}
+
+class MouseTracker extends React.Component {
+  render() {
+    return (
+      <div>
+        <h1>Move the mouse around!</h1>
+        <MouseWithCat/>
+      </div>
+    );
+  }
+}
+```
+
+This approach will work for our specific use case, but we haven't achieved the objective of truly encapsulating the behavior in a reusable way. Now, every time we want the mouse position for a different use case, we have to create a new component (i.e. essentially another `<MouseWithCat>`) that renders something specifically for that use case.
+
+Here's where the render prop comes in: Instead of hard-coding a `<Cat>` inside a `<Mouse>` component, and effectively changing its rendered output, we can provide `<Mouse>` with a function prop that it uses to dynamically determine what to render–a render prop.
+
+```js
+class Cat extends React.Component {
```

Comment:
I'd like to see this written as a SFC to help promote that pattern for simple components like this. I understand that you want to maintain the whole cat/mouse thing, but I believe that one would likely pass `x` and `y` as two props, not as props within a prop, if writing this for real.
```jsx
const Cat = ({ x, y }) => (
  <img src="/cat.jpg" style={{ position: 'absolute', left: x, top: y }}/>
);
```

Reply:
> Passing a mouse prop (as an object with x and y properties) is a leaky abstraction that Cat shouldn't have to worry about.

I personally prefer the `mouse` prop to keep the *semantics* of the code. A real cat doesn't care about x's and y's. But a mouse? Now you've got his attention! ;)

Additionally, passing `{ x, y }` assumes the caller knows too much about what a `<Cat>` *does* (namely, that it chases a mouse and wants to position itself at the mouse's `{ x, y }` coordinates). If `mouse.isHiding` were `true`, the `<Cat>` probably would be completely oblivious to his presence and wouldn't move a muscle.

Reply:
This conversation hits the nail on the head regarding one of the main sticking points of this pattern - enforcing an API contract between the higher-order function and its arguments.

Reply:
Are you kidding me? As you friend @ryanflorence said below:
> the point is to compose Cat and Mouse w/o either having to know about the other

`Cat` is really just what we render at an x/y location. We could pass `Render={SquareRedDiv}`. If we did, you would also likely just pass `x` and `y`. `Cat` or `SquareRedDiv` have no idea that they are trailing a mouse point, nor should they.

Reply:
Forked [CodeSandbox](https://codesandbox.io/s/nnn96mno6p) with `SquareRedDiv` implementation really shows why passing a `mouse` prop is just silly.

```jsx
const style = {
  position: 'absolute', 
  width: 64, 
  height: 64, 
  backgroundColor: 'red',
};

const SquareRedDiv = ({ x, y }) => (
  <div style={{ ...style, left: x, top: y }} />
);
```

Reply:
I honestly don't think it matters very much what the props to `<Cat>` are here. It's just a matter of preference. I like the semantics of `<Cat mouse>` and I like the fact that the caller doesn't need to know any of the details of what the `<Cat>` component is going to do with that information. It's a nice, clean separation in my mind.